### PR TITLE
DAH-1361: Upgrade to React 18

### PR DESCRIPTION
**Pull request by chadbrokaw-exygy**
- Upgrades to React 18
- Removes Webpacker and Webpacker-React in favor of Shakapacker and React-On-Rails, respectively
- Rebuilds build process
- Removes enzyme since it is no longer compatible
- The RTL version of snapshot tests wraps them in a <DocumentFragment>, and renders them closer to actual DOM output so all snapshot tests will have at least the fragment wrapping change + most have lines that changed from className --> class
- Some snapshot tests have a much bigger diff bc they were never actually testing what they were supposed to be testing (some were capturing an empty or loading state because we weren't properly mocking the API)
- All async tests are now using done() which usually isn't necessary, but without them the tests bleed into each other when run as a full suite
- Updating to data-testid
- Adding a few new test ids
- Adding type for type imports - eslint was otherwise v upset
- Upgrades dependencies

### CODE REVIEW

**Comment by jimlin-sfgov** [comment]
The README instructions for spinning up the app locally seem to be outdated now, can you update that so we can manually test this PR locally?

There might be other things that need to be updated in the README, but I haven't gotten far enough in my review.

**Comment by jimlin-sfgov** [comment]
Just making a note here that the branch name does not reflect the work done, and we need to consider that before merging to the main branch.

File: app/controllers/assistance_controller.rb
Comment (jimlin-sfgov) [inline app/controllers/assistance_controller.rb:4]: Just wondering why we're creating all of these instance variables. We get equivalent functionality by calling the `static_asset_paths` helper method inside the view:

```slim
// app/views/assistance/get_assistance.html.slim
== react_component 'GetAssistance', props: { assetPaths: static_asset_paths }
```

Comment (chadbrokaw-exygy) [reply app/controllers/assistance_controller.rb:4]: For sure, we could totally do it that way too! The reason I changed it was because the implementation of just writing `assetPaths: static_asset_paths` wasn't working. My logic for creating the instance variables was that it felt cleaner and easier to read for folks who are new to ruby/react-on-rails. 

Totally happy to change it to your way if you think that is better 😄

Comment (jimlin-sfgov) [reply app/controllers/assistance_controller.rb:4]: I usually prefer to have less code, but if you think it will help others then it may be worth keeping (there's always a balance between terseness and readability). One nit I do have is to use snake_case for variables coming from ruby-land, e.g. `@assistance_props`.

File: app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts
Comment (jimlin-sfgov) [inline app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts:1]: The file that this import is using doesn't exist. It's strange that our linter is not catching this. I'm kind of worried now that there may be other things our linter isn't catching.

File: app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts
Comment (jimlin-sfgov) [inline app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts:3]: Is this data used anywhere? I feel like this file is a work-in-progress, and was forgotten about during the handoff.

File: app/javascript/authentication/SignInForm.tsx
Comment (jimlin-sfgov) [inline app/javascript/authentication/SignInForm.tsx:23]: Is this a note for update work that we have planned?

File: app/javascript/modules/listingDetails/ListingDetailsPricingTable.tsx
Comment (jimlin-sfgov) [inline app/javascript/modules/listingDetails/ListingDetailsPricingTable.tsx:163]: This change doesn't seem related to the React 18 upgrade. I'm just now realizing that this PR combines two unrelated(?) Jira tickets, as indicated by the PR title.

**Review by jimlin-sfgov — COMMENTED** [review]
I wasn't able to review all the changes in one go, here are the questions related to what I've reviewed so far.

File: app/javascript/components/base.scss
Comment (jimlin-sfgov) [inline app/javascript/components/base.scss:111]: Looks like we can remove this class entirely now?

File: app/javascript/hooks/useScript.tsx
Comment (jimlin-sfgov) [inline app/javascript/hooks/useScript.tsx:16]: This is probably a quick fix to appease the linter, but we should try to avoid `any` types in general.

I realize that this file isn't great about typing things, and working to overhaul it is outside the scope of this ticket.

File: app/javascript/hooks/useTranslate.tsx
Comment (jimlin-sfgov) [inline app/javascript/hooks/useTranslate.tsx:54]: See previous comment about `any` types.

File: app/javascript/modules/listingDetailsLottery/ListingDetailsLotterySearchForm.tsx
Comment (jimlin-sfgov) [inline app/javascript/modules/listingDetailsLottery/ListingDetailsLotterySearchForm.tsx:39]: If the `data` param is not guaranteed to have `lotterySearchNumber`, the function should be typed to indicate that.

File: app/javascript/pages/index.tsx
Comment (jimlin-sfgov) [inline app/javascript/pages/index.tsx:39]: Just curious, is this change due to changes with Bloom UIC?

File: config/webpack/base.js
Comment (jimlin-sfgov) [inline config/webpack/base.js:44]: Why do some env vars need `JSON.stringify` while others do not?

File: config/webpack/base.js
Comment (jimlin-sfgov) [inline config/webpack/base.js:51]: I'm sure there is a good reason for this, but it would helpful to add the reason in a comment

**Review by jimlin-sfgov — COMMENTED** [review]
This is part 2 of my review. As mentioned earlier, I had to break it up because I couldn't get through the review in one day.

File: config/initializers/assets.rb
Comment (chadbrokaw-exygy) [reply config/initializers/assets.rb:13]: Will remove!

File: config/shakapacker.yml
Comment (chadbrokaw-exygy) [reply config/shakapacker.yml:1]: Good Catch!

File: app/javascript/pages/index.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/pages/index.tsx:39]: Yes

File: app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts
Comment (chadbrokaw-exygy) [reply app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts:1]: Good catch, and agreed on the linter. I will look into the linter to make sure that it is correctly flagging these types of things.

I wonder if this file was brought in during Mark's work on the pricing table and was forgotten... It doesn't look like it is being used anywhere in the code base (which might be why the linter didn't catch it?). I am going to delete the file unless there is any objection from @emilyjablonski

File: app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts
Comment (chadbrokaw-exygy) [reply app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts:3]: It doesn't look like it is being used anywhere, should be deleted IMO

File: app/javascript/modules/listingDetails/ListingDetailsPricingTable.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/modules/listingDetails/ListingDetailsPricingTable.tsx:163]: @kramduckner Can you comment on this? I think this is a botched merge and should be reverted, but I wanted to confirm

File: app/javascript/modules/listingDetailsLottery/ListingDetailsLotterySearchForm.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/modules/listingDetailsLottery/ListingDetailsLotterySearchForm.tsx:39]: Agreed!

File: config/webpack/base.js
Comment (chadbrokaw-exygy) [reply config/webpack/base.js:51]: I've gone back and forth on this, most [resources](https://github.com/shakacode/react-webpack-rails-tutorial/pull/519) online used it, but the error is not present in our code. 

I have commented out this line and left a comment above it outlining that it could be needed in the future.

File: app/javascript/hooks/useScript.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/hooks/useScript.tsx:16]: Ugh, I completely agree with you. I'm not a fan of any types. This was a quick fix to make the linter happy since we were effectively treating `script` (and below `selectInDom`) as any type anyway. 

I will work on this today to see if I can get the typing right, if I can't then I will create a ticket to tackle it later.

File: config/webpack/base.js
Comment (chadbrokaw-exygy) [reply config/webpack/base.js:44]: I'm not completely sure. I couldn't figure out a justification for that, but this was in the previous config so I decided to leave it as-is.

File: app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts
Comment (emilyjablonski) [reply app/javascript/__tests__/data/RailsListingPricingTableUnits/listing-pricing-table-units-default.ts]: No objection!

**Review by jimlin-sfgov — APPROVED** [review]
My feedback has been addressed, looks good!

I'm approving with the expectation that we will get a few more approvals before merging.

File: app/javascript/hooks/useScript.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/hooks/useScript.tsx:18]: Created a ticket to track this effort: https://sfgovdt.jira.com/browse/DAH-1581

File: app/javascript/hooks/useTranslate.tsx
Comment (chadbrokaw-exygy) [reply app/javascript/hooks/useTranslate.tsx:55]: Created a ticket to track this effort: https://sfgovdt.jira.com/browse/DAH-1581

**Pull request merged**
